gpsgen.py

```diff
@@ -1,9 +1,516 @@
+        self.current_state = next(self.state_machine)
+        return
+
+    def get_position(self):
+        return self.coord
+
+    def __str__(self):
+        return "{city} -> {worker_id} {task_id} {coord}".format(**vars(self))
+
+
+class Step():
+    """
+    1 to 1 wrapper to a Worker, not only for semantics, a Silumation
+    has Steps not Workers, but also for adding the initial step data like
+    start_date.
+    The start_date is the reference date from which next step is processed
+    Worker instance is manipulated from Step too.
+    """
+    def __init__(self, worker,start_date=None):
+        self.worker = worker
+        if start_date is not None:
+            self.time = start_date
+        else:
+            self.time = datetime.datetime.now()
+
+
+    def to_dict(self):
+        return vars(self.worker)
+
+    def full_qualified_id(self):
+        """
+        Full qualified id, that we will use as key on redis(GEOADD).
+        **Note
+        Maybe is a good idea to add a day reference to this, This way web
+        can manage Courier data in redis like a session. Lets see which is
+        the best format for this get_current_day() first class function is
+        for this. Consider also adding this function as a class member.
+        """
+        return "{city}:{worker_id}:{task_id}".format(**self.to_dict())
+
+    def to_json_line(self):
+        return {'id': self.full_qualified_id(),
+                'state': self.worker.current_state.name,
+                'coord': [self.worker.coord.lat, self.worker.coord.lon],
+                'timestamp': float(time.mktime(self.time.timetuple()))
+                }
+
+    def __str__(self):
+        step_data  =  (self.full_qualified_id(),
+                       self.worker.current_state,
+                       self.worker.coord,
+                       self.time)
+        f_string = "[<STEP> for {}] [<ACTION>:{}] [<COORD>:{}] [<TIME>: {}]"
+        return f_string.format(*step_data)
+
+
+class StepScheduler():
+    def __init__(self):
+        self._step_queue = deque()
+
+    def new_step(self, step):
+        self._step_queue.append(step)
+
+    def run(self):
+        while self._step_queue:
+            step = self._step_queue.popleft()
+            try:
+                next(step)
+                self._step_queue.append(step)
+            except StopIteration:
+                pass
+
+def io_coroutine(gen):
+    """
+    This decorator, pumps the coroutine to the next step.
+    It is also a Syntactic sugar for identifying coroutines.
+    Which are much more visible now.
+    Registers the initiliced coroutine for closing them when
+    the data flow ends, this is specially important, when
+    opened file descriptors are involved, in this case we are
+    sending data to a "with open" block, registering coroutines
+    allow us to kill them easyly with Simulator.end_io_coroutines()
+    """
+    def pumped(self, *args, **kwargs):
+        g = gen(self, *args, **kwargs)
```

Comment:
It's good practise to use `@functools.wraps` to preserve the name and docstring of the function you're decorating.